core/src/main/java/org/infinispan/interceptors/locking/PessimisticLockingInterceptor.java

```diff
@@ -190,95 +145,89 @@ public Object visitLockControlCommand(TxInvocationContext ctx, LockControlComman
-      if (skipLocking) {
+      if (command.isUnlock()) { //TODO double check unlock. it isn't used anymore!
+         if (ctx.isOriginLocal()) {
+            throw new AssertionError(
+                  "There's no advancedCache.unlock so this must have originated remotely.");
+         }
          return false;
       }
 
-      // First go through the distribution interceptor to acquire the remote lock - required by DLD.
-      // Only acquire remote lock if multiple keys or the single key primary owner is not the local node.
-      if (ctx.isOriginLocal()) {
-         final boolean isSingleKeyAndLocal =
-               !command.multipleKeys() && cdl.getCacheTopology().getDistribution(command.getSingleKey()).isPrimary();
-         boolean needBackupLocks = !isSingleKeyAndLocal || isStateTransferInProgress();
-         if (needBackupLocks && !command.hasAnyFlag(FlagBitSets.CACHE_MODE_LOCAL)) {
-            LocalTransaction localTx = (LocalTransaction) ctx.getCacheTransaction();
-            if (localTx.getAffectedKeys().containsAll(command.getKeys())) {
-               if (trace)
-                  log.tracef("Already own locks on keys: %s, skipping remote call", command.getKeys());
-               return true;
-            }
-         } else {
-            if (trace)
-               log.tracef("Single key %s and local, skipping remote call", command.getSingleKey());
-            return localLockCommandWork(ctx, command);
+      if (hasSkipLocking(command) || areAllKeysAlreadyLocked(ctx, command.getKeys())) {
+         //lock control with skip locking?? that is weird...
+         if (trace) {
+            log.tracef("Skip locking or keys already locked. Skipping command %s", command);
          }
+         return false;
+      } else if (isLocalOnly(command)) {
+         lockOrRegisterBackupLock(ctx, command.getKeys(), getLockTimeoutMillis(command));
+         ctx.addAffectedKey(command.getKeys());
+         return invokeNext(ctx, command);
+      } else {
+         return ctx.isOriginLocal() ?
+                handleLocalLockCommand(ctx, command) :
+                handleRemoteLockCommand(ctx, command);
       }
+   }
 
-      return invokeNextThenApply(ctx, command, localLockCommandWork);
+   private boolean readNeedsLock(InvocationContext ctx, FlagAffectedCommand command) {
+      return ctx.isInTxScope() &&
+             ctx.isOriginLocal() &&
+             command.hasAnyFlag(FlagBitSets.FORCE_WRITE_LOCK) &&
+             !hasSkipLocking(command);
    }
 
-   private boolean localLockCommandWork(InvocationContext ctx, LockControlCommand command)
-         throws InterruptedException {
-      TxInvocationContext<?> txInvocationContext = (TxInvocationContext<?>) ctx;
-      if (ctx.isOriginLocal()) {
-         txInvocationContext.addAllAffectedKeys(command.getKeys());
-      }
+   private boolean areAllKeysAlreadyLocked(TxInvocationContext<?> context, Collection<?> keys) {
+      return context.getAffectedKeys().containsAll(keys);
```

Comment:
If you execute put with skip locking, and then putMap with the same key, the key is already in the context but it's not locked... amiright?

Reply:
I've noticed that there is a `InvocationContext.hasLockedKey` method.

Reply:
`hasLockedKey` only check if this node has the primary lock acquired. this is false when the transaction is running in the backup owner or a non-owner. On other hand, the affected keys contains all the touched keys.